control-plane/config-entries/controllersv2/mesh_gateway_controller.go

```diff
@@ -69,11 +72,83 @@ func (r *MeshGatewayController) SetupWithManager(mgr ctrl.Manager) error {
 }
 
 func (r *MeshGatewayController) onCreateUpdate(ctx context.Context, req ctrl.Request, resource *meshv2beta1.MeshGateway) error {
-	// TODO NET-6392 NET-6393 NET-6394 NET-6395
-	return errors.New("onCreateUpdate not implemented")
+	builder := gateways.NewMeshGatewayBuilder(resource)
+
+	ifNew := func(ctx context.Context, object client.Object) error {
+		return r.Create(ctx, object)
+	}
+
+	ifExists := func(ctx context.Context, object client.Object) error {
+		return r.Update(ctx, object)
```

Comment:
Just wondering, but would we need to take into account things like managed kubernetes service annotation additions or anything that mutates some sort of metadata here? Not sure if any implementations mutate `ServiceAccount` objects or not... but I know there's some stuff added to `Service` objects by EKS and (I believe?) GKE that we need to ignore to avoid infinite reconciliation problems.

Accordingly, just wondering if maybe it's worthwhile to make the `ifExists` closure here take an `oldEntry, newEntry client.Object` argument so we can possibly do some merging of resources on an update?

Reply:
Good callout! I believe we only do this merging logic for [Service](https://github.com/hashicorp/consul-k8s/blob/71cdbc24fa8ad3bf3571b4c71744a5171cafdd50/control-plane/api-gateway/gatekeeper/service.go#L37-L43) and [Deployment](https://github.com/hashicorp/consul-k8s/blob/085c812761439c0a1be774dc7158eb441901099f/control-plane/api-gateway/gatekeeper/deployment.go#L58-L64) today.

Tomorrow, I'll make sure we have a path forward on that. Might wait to actually introduce that code in the PR where it's needed.

Reply:
Wound up modifying the operation func to receive the old object and the new object to allow for merging for other resources @andrewstucki